internal/bufio/bufio.go

```diff
@@ -0,0 +1,205 @@
+// Package bufio provides specialised packet buffers.
+package bufio
+
+import (
+	"bytes"
+	"errors"
+	"io"
+)
+
+// ErrNegativeCount is returned when a negative number of bytes is requested.
+var ErrNegativeCount = errors.New("bufio: negative count")
+
+// PacketReader implements buffering for an io.Reader object.
+// PacketReader will not read from the underlying reader unless its
+// buffer is empty, reducing the risk of cross packet reading.
+type PacketReader struct {
+	rd   io.Reader
+	buf  []byte
+	r, w int
+}
+
+// NewPacketReader returns a new PacketReader whose buffer is size.
+func NewPacketReader(rd io.Reader, size int) *PacketReader {
+	r := &PacketReader{}
+	r.reset(make([]byte, size), rd)
+	return r
+}
+
+// Reset discards any buffered data, resets all state, and switches
+// the buffered reader to read from rd.
+// Calling Reset on the zero value of Reader will panic.
+func (r *PacketReader) Reset(rd io.Reader) {
+	if r.buf == nil {
+		panic(errors.New("bufio: uninitialized PacketBuffer"))
+	}
```

Comment:
Why not just do nothing instead of panicking? Resetting the zero value could just do nothing, I guess. WDYT? It's only an internal package for now but it's usually better not to panic when we can avoid it.